cartographer_rviz/cartographer_rviz/drawable_submap.cc

```diff
@@ -159,13 +168,16 @@ void DrawableSubmap::SetAlpha(const double current_tracking_z) {
 
 void DrawableSubmap::UpdateSceneNode() {
   ::cartographer::common::MutexLocker locker(&mutex_);
+  render_in_progress_ = false;
+  if (quitting_) {
```

Comment:
TL;DR: I think the issue can be fixed by replacing this commit by just adding to the beginning of the destructor:
```
if (query_in_progress_) {
  rpc_request_future_.wait();
}
```
Does this solve the observed issues for you?

You have to be more specific for me to follow your reasoning. For example, you set `quitting_` to true in the destructor, but use it in `MaybeFetchTexture()` line 125. Why would this member function be called if the object was destroyed? It seems this should and does never happen. Line 172 seems to defend against the fact that the object to which `this` points was already destroyed. At this point looking at `quitting_` is too late as the object is already gone.

My investigation: In the problematic case, the main thread in `submaps_display.cc` calls `trajectories_.clear()` while a query is in progress. We need to wait before we destroy the `DrawableSubmap` object. Waiting for `query_in_progress_` to become false is not possible, since it causes a deadlock: The destructor runs in the main thread, so blocks Qt from handling signals. If instead we just wait until the signal is emitted (that's my code suggestion above), Qt knows not to call a slot if it was deleted after emitting as far as I understand.